earthfile2llb/withdockerrun.go

```diff
@@ -341,14 +341,14 @@ func (wdr *withDockerRun) solveImage(ctx context.Context, mts *states.MultiTarge
 	sessionID := hex.EncodeToString(sha256SessionIDKey[:])
 	// Add the tar to the local context.
 	tarContext = llb.Local(
-		solveID,
+		keyFunc(),
 		llb.SessionID(sessionID),
 		llb.Platform(llbutil.DefaultPlatform()),
 		llb.WithCustomNamef("[internal] docker tar context %s %s", opName, sessionID),
 	)
 	wdr.tarLoads = append(wdr.tarLoads, tarContext)
-	wdr.c.mts.Final.LocalDirs[solveID] = outDir
-	wdr.c.opt.SolveCache[solveID] = tarContext
+	wdr.c.mts.Final.LocalDirs[keyFunc()] = outDir
+	wdr.c.opt.SolveCache.Set(keyFunc, tarContext)
```

Comment:
More or less the problem was what the target hash is not enough here to make assumptions on. We need to also include the name its being assigned.